### PR TITLE
Fallacious slice: First pass at server side rendering

### PR DESCRIPTION
**Pull request by GregWeil**
## Links
- https://fallacious-slice.glitch.me/
- https://app.zenhub.com/workspaces/community-5d07d1359dee6059a9688176/issues/fogcreek/glitch-community-work/188

## Changes:
- The homepage and create page both get rendered on the server! Other pages are kept on the client because they would render as just a loading spinner which would (I think) trample the seo stuff
- Introduced babel-register to translate js imported from src/ so it works in node
- Using chokidar to watch the src/ directory and force them to be reimported on next use
(aka hot reloading, because restarting node would also restart the webpack build)
- Several smaller changes around the site as I found stuff that was broken

## How To Test:
- Try refreshing the page around the site. The homepage and create page should be quite a bit faster
(I have it set to production mode so you can do comparisons with community.glitch.me)
- Try loading pages with IE/js disabled/in google translate. They could do some more fancy stuff, but for now I'm keeping it simple and trying to keep the behavior unchanged

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,7 +18,7 @@
   ],
   "scripts": {
     "prestart": "nodemon --watch server --watch shared --exec 'eslint --config server/.eslintrc.server.js server shared webpack.config.js' &",
-    "start": "nodemon --watch server --watch shared --watch views --exec 'forever --minUptime 5000 --spinSleepTime 5000' ./server/server.js",
+    "start": "forever --minUptime 5000 --spinSleepTime 1000 -c 'nodemon --watch server --watch shared --watch views --exitcrash' ./server/server.js",
```
Comment (GregWeil) [inline package.json]: I flipped the order here, which fixed a bunch of 'port in use' errors I was getting when editing server code

**Comment by taravancil** [comment]
This is awesome, I'm excited. I'll take a look either tonight or tomorrow.

File: src/components/header/logo.js [hunk 2]
```diff
@@ -9,8 +9,10 @@ const LOGO_NIGHT = 'https://cdn.glitch.com/2bdfb3f8-05ef-4035-a06e-2043962a3a13%
 const LogoImage = ({ src }) => <img className={styles.logo} src={src} alt="Glitch" />;
 
 function Logo() {
-  const [hour, setHour] = useState(new Date().getHours());
+  const [hour, setHour] = useState(12);
```
Comment (GregWeil) [inline src/components/header/logo.js]: SSR-ed pages always show the day fish, then the browser immediately corrects it.

File: src/presenters/pages/router.js [hunk 3]
```diff
@@ -79,110 +79,113 @@ const PageChangeHandler = withRouter(({ location }) => {
...
-);
+        {EXTERNAL_ROUTES.map((route) => (
+          <Route key={route} path={route} render={({ location }) => <ExternalPageReloader key={location.key} />} />
+        ))}
+
+        <Route render={({ location }) => <NotFoundPage key={location.key} />} />
+      </Switch>
+    </>
+  );
+};
```
Comment (GregWeil) [inline src/presenters/pages/router.js]: This is mostly whitespace changes after the addition of `useGlobals()`

File: src/utils/sentry.js [hunk 4]
```diff
@@ -9,61 +9,64 @@
...
 
-  // Expose for use on the developer console:
-  window.Sentry = Sentry;
-} catch (error) {
-  console.warn('Error initializing Sentry', error);
+    // Expose for use on the developer console:
+    window.Sentry = Sentry;
+  } catch (error) {
+    console.warn('Error initializing Sentry', error);
+  }
```
Comment (GregWeil) [inline src/utils/sentry.js]: Also just whitespace changes to put the whole hing in an if statement

**Comment by taravancil** [comment]
I want to give this one final review before I approve it, but it all looks great!

Some metrics for comparison (lower is better for each one):
<table>
<tr>
  <th></th>
  <th>community homepage</th>
  <th>SSR homepage</th>
  <th>community /create</th>
  <th>SSR /create</th>
</tr>

<tr>
 <td>first contentful paint</td>
  <td>6.0.s</td>
  <td>1.7s 🎉</td>
  <td>1.8s</td>
  <td>1.6s</td>
</tr>

<tr>
  <td>first meaningful paint</td>
  <td>6.6.s</td>
  <td>2.9s 🎉</td>
  <td>4.8s</td>
  <td>2.3s 🎉</td>
</tr>

<tr>
  <td>speed index</td>
  <td>7.1s</td>
  <td>10.0s</td>
  <td>5.7s</td>
  <td>4.4s</td>
</tr>
</table>

The improvements to first contentful/meaningful paint are great and to be expected. Time to interactive was not improved and that makes sense too. The client still needs to load and parse the bundle.

**Pull request merged**